src/LeaseManager.h

```diff
@@ -0,0 +1,157 @@
+ * module.  Servers in their part must contact this module to check the validity
+ * of client leases.
+ */
+class LeaseManager {
+  PUBLIC:
+    explicit LeaseManager(Context *context);
+    WireFormat::ClientLease getLeaseInfo(uint64_t leaseId);
+    void recover();
+    WireFormat::ClientLease renewLease(uint64_t leaseId);
+    void startUpdaters();
+
+  PRIVATE:
+    /**
+     * The LeasePreallocator is periodically invoked to allocate a leaseId on
+     * external storage.  The goal is that this preallocator will work ahead
+     * of the issued leases so that a client does not have to wait for an
+     * external storage operation to complete a new lease request.  Every
+     * invocation of the preallocator should ensure that maxAllocatedLeaseId
+     * runs ahead of lastIssuedLeaseId by the PREALLOCATION_LIMIT.  This batch
+     * allocation process only blocks during each individual allocation; other
+     * operations like issuing leases can be safely interleaved.
+     */
+    class LeasePreallocator : public WorkerTimer {
+      public:
+        explicit LeasePreallocator(Context* context,
+                                   LeaseManager* leaseManager);
+        virtual void handleTimerEvent();
+
+        LeaseManager* leaseManager;
+      private:
+        DISALLOW_COPY_AND_ASSIGN(LeasePreallocator);
+    };
+
+    /**
+     * The LeaseCleaner periodically wakes up to expire leases.  Every time it
+     * does a cleaning pass, it will attempt to clean as many leases as
+     * possible.  The cleaning process only blocks during the cleaning of a
+     * single lease; other operations like lease renewal can safely interleave
+     * with cleaning during a cleaning pass.
+     */
+    class LeaseCleaner : public WorkerTimer {
+      public:
+        explicit LeaseCleaner(Context* context,
+                              LeaseManager* leaseManager);
+        virtual void handleTimerEvent();
+
+        LeaseManager* leaseManager;
+      private:
+        DISALLOW_COPY_AND_ASSIGN(LeaseCleaner);
+    };
+
+    /// Monitor-style lock
+    SpinLock mutex;
+    typedef std::lock_guard<SpinLock> Lock;
+
+    /// Shared information about the server.
+    Context* context;
+
+    /// Used to provided a guaranteed monotonically increasing number (even
+    /// across crashes) that (mostly) advances in sync with real time.  This
+    /// Cluster Time is used to define lease expiration times.
+    CoordinatorClusterClock clock;
+
+    /// Represents the largest leaseId that was issued from this module.  The
+    /// module guarantees that no leaseId will be issued more than once.  The
+    /// next leaseId issued should be ++lastIssuedLeaseId.
+    uint64_t lastIssuedLeaseId;
+
+    /// This is the largest leaseId that has been pre-allocated in external
+    /// storage.  Pre-allocating leaseIds allows this module to respond to
+    /// requests for new leases without waiting for an external storage
```

Comment:
This documentation appears to duplicate  some of the documentation around line 50 above. Figure out a way to reduce the duplication.